### PR TITLE
[Hardware][Neuron] Refactor neuron support

### PR DESCRIPTION
**Pull request by zhuohan123**
Refactor the neuron support. After this refactoring, the neuron support is completely isolated from the GPU support. The development on the GPU side no longer needs to consider its effect on the neuron pass. Also, all the `is_neuron` flags in the code have been removed, except the one that tells us that we are using `neuron` device. 

Highlights of this change:

- Isolated `NeuronExecutor`, `NeuronWorker`, `NeuronModelRunner`, and `get_neuron_model` function.
- Remove all the unnecessary neuron model definitions. Have a general abstraction of neuron models.

Remaining issues for future PRs:
- The `NeuronModelRunner` still has a complicated `prepare_inputs` function. This should be refactored together with the refactor of `prepare_inputs` in the future.
- We probably need more tests for the neuron pass. I'll leave this task to the AWS team (cc @liangfu).

---

<details>

<summary><b> PR Checklist (Click to expand. Please read before submitting.) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by zhuohan123** [comment]
@WoosukKwon This PR is ready for review!

cc @liangfu please take a look as well!

File: examples/offline_inference_neuron.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 # Create an LLM.
 llm = LLM(
-    model="openlm-research/open_llama_3b",
+    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
```
Comment (zhuohan123) [inline examples/offline_inference_neuron.py:15]: This model is smaller and can run on the smallest neuron instance.

Comment (liangfu) [reply examples/offline_inference_neuron.py:15]: nice!

File: vllm/config.py [hunk 2]
```diff
@@ -408,24 +408,15 @@ def __init__(
...
-        if is_neuron():
-            # For Neuron device support, here we assign TP=1 to avoid sharding
-            # within vLLM directly. Transformer-neuronx would take
-            # neuron_tp_degree attribute, and distribute the workload
-            # to multiple NeuronCores.
-            self.tensor_parallel_size = 1
-            self.neuron_tp_degree = tensor_parallel_size
-        else:
-            self.tensor_parallel_size = tensor_parallel_size
+        self.tensor_parallel_size = tensor_parallel_size
```
Comment (liangfu) [inline vllm/config.py:411]: Have you tested with tensor_parallel_size>1 on neuron device?

File: vllm/engine/async_llm_engine.py [hunk 3]
```diff
@@ -325,7 +325,12 @@ def from_engine_args(cls,
         # Create the engine configs.
         engine_configs = engine_args.create_engine_configs()
         parallel_config = engine_configs[2]
-        if parallel_config.worker_use_ray or engine_args.engine_use_ray:
+        device_config = engine_configs[4]
+
+        if device_config.device_type == "neuron":
```
Comment (liangfu) [inline vllm/engine/async_llm_engine.py:330]: Instead of string comparison, set a `property` in device_config?

FYI, we might consider some device-specific config with `device="neuron config=(a=1, b=2, c=3)"`

File: vllm/model_executor/neuron_model_loader.py [hunk 4]
```diff
@@ -20,31 +25,88 @@
...
+        super().__init__()
+        self.config = config
+        self.model = None
+        self.sampler = Sampler(config.vocab_size, logits_as_input=True)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        input_block_ids: torch.Tensor,
```
Comment (liangfu) [inline vllm/model_executor/neuron_model_loader.py:52]: I still prefer passing input_metadata into the forward call, in order to support paged attention. But feel free to chime in.

**Review by liangfu — CHANGES_REQUESTED** [review]
Thanks @zhuohan123 for the proposed change. I like the idea of
* abstracting execution with `executor_class`
* deduplicating llama and mixtral model support

My concern comes from missing input_metadata in the forward function call, where we might need an extra refactoring to support paged attention on neuron backend.

**Comment by liangfu** [comment]
Tested with tensor_parallel_size=2 for the branch:
```
INFO 03-19 22:26:56 config.py:446] Custom all-reduce kernels are temporarily disabled due to stability issues. We will re-enable them once the issues are resolved.
INFO 03-19 22:26:56 llm_engine.py:67] Initializing an LLM engine (v0.3.3) with config: model='TinyLlama/TinyLlama-1.1B-Chat-v1.0', tokenizer='TinyLlama/TinyLlama-1.1B-Chat-v1.0', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=128, download_di\
r=None, load_format=auto, tensor_parallel_size=2, disable_custom_all_reduce=True, quantization=None, enforce_eager=False, kv_cache_dtype=auto, device_config=cpu, seed=0)
tokenizer_config.json: 100% 1.29k/1.29k [00:00<00:00, 324kB/s]
tokenizer.model: 100% 500k/500k [00:00<00:00, 12.1MB/s]
tokenizer.json: 100% 1.84M/1.84M [00:00<00:00, 26.8MB/s]
special_tokens_map.json: 100% 551/551 [00:00<00:00, 1.14MB/s]
WARNING 03-19 22:26:57 utils.py:324] Pin memory is not supported on Neuron.
model.safetensors: 100% 2.20G/2.20G [00:08<00:00, 260MB/s]
generation_config.json: 100% 124/124 [00:00<00:00, 33.9kB/s]
2024-03-19 22:27:23.000450:  1486537  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
2024-03-19 22:27:23.000452:  1486537  INFO ||NEURON_CC_WRAPPER||: Call compiler with cmd: neuronx-cc compile --target=trn1 --framework=XLA /tmp/ubuntu/neuroncc_compile_workdir/facc01e3-5e6a-4cf8-9d45-a1bca66c03ec/model.MODULE_622c52b3b2f99196d008+2c2d707e.hlo_module.pb --output /tmp/ubuntu/neuroncc_compile_workdir\
/facc01e3-5e6a-4cf8-9d45-a1bca66c03ec/model.MODULE_622c52b3b2f99196d008+2c2d707e.neff --model-type=transformer --auto-cast=none --verbose=35
2024-03-19 22:27:23.000485:  1486538  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
2024-03-19 22:27:23.000487:  1486538  INFO ||NEURON_CC_WRAPPER||: Call compiler with cmd: neuronx-cc compile --target=trn1 --framework=XLA /tmp/ubuntu/neuroncc_compile_workdir/73dc4908-e815-4c42-8313-5bd0daa701a6/model.MODULE_067094cf9b96d1dd4d8c+2c2d707e.hlo_module.pb --output /tmp/ubuntu/neuroncc_compile_workdir\
/73dc4908-e815-4c42-8313-5bd0daa701a6/model.MODULE_067094cf9b96d1dd4d8c+2c2d707e.neff --model-type=transformer --auto-cast=none --verbose=35
........
Compiler status PASS
2024-03-19 22:28:33.000124:  1486538  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
.
Compiler status PASS
2024-03-19 22:28:45.000561:  1486537  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
2024-Mar-19 22:28:45.0977 1486374:1486511 [1] nccl_net_ofi_init:1415 CCOM WARN NET/OFI aws-ofi-nccl initialization failed
2024-Mar-19 22:28:45.0977 1486374:1486511 [1] init.cc:137 CCOM WARN OFI plugin initNet() failed is EFA enabled?
Processed prompts: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 16.53it/s]
Prompt: 'Hello, my name is', Generated text: ' _____. I am a ________. I am passionate about _'
Prompt: 'The president of the United States is', Generated text: ' a man named Donald Trump. 4. Biden/Harris team'
Prompt: 'The capital of France is', Generated text: ' Paris. France has a total area of 552,655'
Prompt: 'The future of AI is', Generated text: ' in collaboration. Here are four examples:\n\n1. IBM Watson: This'
```

File: vllm/config.py (hunk 2 above)
Comment (zhuohan123) [reply vllm/config.py:411]: Tested by liangfu and it works great!

File: vllm/engine/async_llm_engine.py (hunk 3 above)
Comment (zhuohan123) [reply vllm/engine/async_llm_engine.py:330]: `device_type` is different from `device`. I think we can safely assume `device_type` can only be `cuda` and `neuron` for now? We can put the config string in other fields of `device_config`.

File: vllm/model_executor/neuron_model_loader.py [hunk 5]
```diff
@@ -20,31 +25,88 @@
     torch.float32: "f32",
 }
 
+# Models supported by Neuron.
+_NEURON_SUPPORTED_MODELS = {
+    "LlamaForCausalLM": ("transformers_neuronx.llama.model",
+                         "LlamaForSampling", "LlamaForCausalLM"),
+    "MistralForCausalLM": ("transformers_neuronx.mistral.model",
+                           "LlamaForSampling", "MistralForCausalLM")
```
Comment (timelfrink) [inline vllm/model_executor/neuron_model_loader.py:33]:
```suggestion
                           "MistralForSampling", "MistralForCausalLM")
```
Mistral has own sampler in transformers_neuronx

Comment (zhuohan123) [reply vllm/model_executor/neuron_model_loader.py:33]: thanks for the great catch!

File: vllm/engine/async_llm_engine.py (hunk 3 above)
Comment (liangfu) [reply vllm/engine/async_llm_engine.py:330]: Sure, i think we are fine to move on with this.

**Review by liangfu — APPROVED** [review]
LGTM

File: vllm/worker/neuron_model_runner.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        parallel_config: ParallelConfig,
+        scheduler_config: SchedulerConfig,
+        device_config: DeviceConfig,
+    ):
+        self.model_config = model_config
+        self.parallel_config = parallel_config
+        self.scheduler_config = scheduler_config
+
+        if model_config is not None and model_config.get_sliding_window():
+            raise ValueError("Sliding window is not supported on Neuron.")
```
Comment (timelfrink) [inline vllm/worker/neuron_model_runner.py:34]: Is this correct? For example mistral is supported by neuron which has `sliding_window` in model config.

Comment (zhuohan123) [reply vllm/worker/neuron_model_runner.py:34]: thanks! change error to a warning.

File: vllm/utils.py [hunk 7]
```diff
@@ -338,7 +338,30 @@ def create_kv_caches_with_random(
...
+        print_warning_once("Using 'pin_memory=False' as WSL is detected. "
+                           "This may slow down the performance.")
+        return False
+    elif is_neuron():
+        print_warning_once("Pin memory is not supported on Neuron.")
+        return False
+    return True
+
+
+class CudaMemoryMeasurer:
```
Comment (WoosukKwon) [inline vllm/utils.py:364]: nit: What about `CudaMemoryProfiler` or `CudaMemoryTracker` or something like that?

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for the great work! This is great refactoring. The code looks much cleaner now.

File: vllm/model_executor/__init__.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 from vllm.model_executor.input_metadata import InputMetadata
 from vllm.model_executor.sampling_metadata import SamplingMetadata
-from vllm.model_executor.utils import set_random_seed, get_model
```
[the remark this answers is not in the document]
Comment (zhuohan123) [reply vllm/model_executor/__init__.py:3]: there are two `get_model`s, one for neuron and one for GPUs. Having `get_model` here will be confusing.

[the remark this answers is not in the document]
Comment (zhuohan123) [reply vllm/model_executor/__init__.py:3]: Also we removed the `get_model` function from the utils. Previously this `get_model` is a dispatcher for different devices. Now we no longer need the dispatcher.

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/__init__.py:3]: Then should we remove it from __all__ as well?

[the remark this answers is not in the document]
Comment (zhuohan123) [reply vllm/model_executor/__init__.py:3]: yeah it's already removed. See the latest commit

File: vllm/utils.py [hunk 9]
```diff
@@ -338,7 +338,26 @@ def create_kv_caches_with_random(
     return key_caches, value_caches
 
 
-class measure_cuda_memory:
+@lru_cache
+def print_warning_once(msg: str) -> None:
+    logger.warning(msg)
+
+
+def is_pin_memory_available() -> bool:
```
Comment (WoosukKwon) [inline vllm/utils.py:346]: QQ: Does it make sense to use cache for this function? Wondering if it helps performance.

Comment (zhuohan123) [reply vllm/utils.py:346]: yes let me add that
